### PR TITLE
Fix useful_to_s? on older Rubies < 2.0

### PR DESCRIPTION
**Pull request by sfcgeorge**
`useful_to_s?` wasn't useful on Ruby < 2.0 as `#inspect` simply called `#to_s` so they were always the same. I've fixed that so it works properly in all Rubies.

### CODE REVIEW

File: lib/contracts/formatters.rb [hunk 1]
```diff
@@ -85,8 +85,12 @@ def plain?
 
       def useful_to_s?
         # Useless to_s value or no custom to_s behavious defined
-        # Ruby < 2.0 makes inspect call to_s so this won't work
-        @value.to_s != '' && @value.to_s != @value.inspect
+        @value.to_s != '' &&
+          if @value.class == Class # It's a class contract
```
Comment (waterlink) [inline lib/contracts/formatters.rb]: multiline `if..else..end` as a second condition to complex condition after `&&` looks confusing to me. Any chance extracting `if..else..end` part to method?

Comment (sfcgeorge) [inline lib/contracts/formatters.rb]: Hah you're right, I'll see what I can do.

**Comment by sfcgeorge** [comment]
Phew, that was a pain to fix. I'm still not convinced it's entirely correct but at least all current tests and the one I added are passing. 

inspect calls to_s in old Rubies so we can't compare them to see if to_s is useful. Instead, if it's a class I check to see if to_s is the same as the class name, if it's an object I check to see if to_s looks like an object representation `#<Class:...>`.

But then if it wasn't useful... we were calling inspect, which in those old Rubies did the same thing as to_s, so we were substituting a useless to_s call with a useless inspect call. I wrote a raw_inspect that tries to do something better. If to_s isn't empty then inspect is called and the result will vary between Rubies. If to_s is empty (specific case I was trying to fix) then it calls name instead. This will look right for classes but won't look right for objects because it isn't object representation. I'm open to changes of behaviour in this method. The basic requirement is that it doesn't print an empty string if to_s is empty string.

**Comment by waterlink** [comment]
LGTM

File: lib/contracts/formatters.rb [hunk 2]
```diff
@@ -85,8 +85,28 @@ def plain?
...
-        @value.to_s != "" && @value.to_s != @value.inspect
+        !empty_to_s? && custom_to_s?
+      end
+
+      def empty_to_s?
+        @value.to_s.empty?
+      end
+
+      def custom_to_s?
+        if @value.class == Class # It's a class contract
```
Comment (waterlink) [inline lib/contracts/formatters.rb]:
```ruby
if @value.class == Class
  # ...
else
  # ...
end
```

This construction duplicated. Not that it is so bad, but could be better:

```ruby
class InspectWrapper
  def lifted_class
    @value.class == Class ? ClassContractInspectWrapper : InstanceContractInspectWrapper
  end

  def lift
    lifted_class.new(@value)
  end

  def custom_to_s?
    lift.custom_to_s?
  end

  def raw_inspect
    lift.raw_inspect.gsub(/^Contracts::/, "")
  end
end

class InstanceContractInspectWrapper
  def custom_to_s?
    !@value.to_s.match(/#\<\w+:.+\>/)
  end

  def raw_inspect
    empty_to_s? ? @value.class.name : @value.inspect
  end
end

class ClassContractInspectWrapper
  def custom_to_s?
    @value.to_s != @value.name
  end

  def raw_inspect
    empty_to_s? ? @value.name : @value.inspect
  end
end
```

More complicated, but more DRY and robust.

Comment (waterlink) [inline lib/contracts/formatters.rb]: Don't really know what is better in that case. So for you @sfcgeorge to decide :)

**Comment by sfcgeorge** [comment]
To reduce duplication and allow for speciffic class vs object inspect behaviour I did something similar to what @waterlink suggested and turned InspectWrapper into a factory. I wondered if `self.new` can be overridden, and it can, it just requires a little alias trick. I love this, I think it's equal parts genius and madness. If you think it's too far on the mad side I can do something more sane though.

**Comment by alex-fedorov** [comment]
Looks good to me, WDYT @egonSchiele on this?

**Comment by egonSchiele** [comment]
I don't like overriding `new`. I would just make a method on `InspectWrapper` called `build` or something. Less patching == less chance of running into obscure edge cases.

**Comment by sfcgeorge** [comment]
I found a simpler less magic way to override `new`. I like this abstract factory approach best because it leaves the implementation of InspectWrapper hidden, and you don't have to keep track of which objects have to be instantiated with `new` and which with `create`.

If you don't like the use of `allocate` then I can change InspectWrapper to a module and then there's no problem defining `new` on it, then `ClassInspectWrapper` and `ObjectInspectWrapper` would include it rather than inheriting from it.

Or if you really don't like overriding `new` however it's done then I can change it to `create`.

**Comment by egonSchiele** [comment]
I don't like the idea of overriding `new`. My issues are:

- Unexpected behavior: When I call `InspectWrapper.new`, I expect to get an instance of `InspectWrapper`.
- Unnecessary complexity. I don't know all the consequences that this has. It's fine for the use case we have right now. But maybe down the line we will add more functionality to the wrapper classes, and run into a subtle bug with this. I don't want to have to deal with that.

I think we should make it explicit that `InspectWrapper` is a factory, so I think using `create` instead of `new` is a good thing.

One thing I like about this implementation: it means `InspectWrapper` is totally abstract, you can't accidentally make an instance of `InspectWrapper`. But I don't think it's worth trading off the issues I listed.

**Comment by sfcgeorge** [comment]
Ok I'll change to `create`, and fix the Rubocop mistakes while I'm at it.

**Comment by sfcgeorge** [comment]
I've removed the overriding `new` in favour of `create`. I've seen the create naming used in other gems.

I have made InspectWrapper a module and given it a class method `new` that raises an exception explaining that it's a factory and `create` must be used. It helped me in changing all the old calls to `new`, if you don't think it's necessary I can quickly remove it.

**Comment by nixpulvis** [comment]
I'm a bit disconnected from this discussion, however seems like a module would have fit the bill better than this "abstract class" since Ruby doesn't really have abstract classes.

**Comment by sfcgeorge** [comment]
Not strict ones no, it's conceptually abstract :) But yeah I suppose that is a reason to use a module, good thought. 

That said, a fake `new` method is now unnecessary as modules can't be instantiated so I've removed it. I did the change in 2 steps and hadn't fully thought it through. Simpler now.

**Comment by nixpulvis** [comment]
I'm not necessarily saying you should rewrite it (not my place), but just a thought for your future Ruby adventures.

**Comment by sfcgeorge** [comment]
Ugh I broke it, sorry for all the revisions. I've fixed it and rebase squashed the last few commits together.

**Comment by egonSchiele** [comment]
thanks!

**Pull request merged**